### ISSUE TITLE
1-2 seconds server lag when someone tries to report someone offline

### ISSUE THREAD

**Issue by mokdennie26** [MrTigreroux/TigerReports#77]
Hi.

When someone tries to `/report name`, and if that `name` is an offline player or someone that doesn't even exist, then the server will lag for like 1 second trying to find the user data.

Well, this issue exists for a long time, even with the following settings set to correct value in config.

> ReportOnline: yes

I just got the time to report it since someone can easily exploit this to lag the server by spamming this command over and over again on an offline player.

For example, if you type `/report wefwwewr`,
Then before the message `Player wefwwer has never joined the server` shows up, the server stuttered. I believe that's due to the fact that the plugin is searching for the corresponding user data regardless of the ReportOnline settings. Maybe it's because of some errors with the order of the related codes. The plugin should check the value of ReportOnline first, then decide whether to perform a user data search.

In other words, it will be better if the plugin skips looking up for offline user data if ReportOnline is set to yes.

That's why I recommend a fix to the ReportOnline settings. It should completely block players from reporting anyone who is offline / not exists.

> TigerReports Version: 4.12.10
> Spigot: Vanilla Spigot 1.8.8
> Online-mode: false

Thanks for the excellent plugin. Everything else works fine.

**Comment by MrTigreroux** [comment]
Hello,
Thank you for reporting the issue. I think that the problems comes from a check that is done to see if the user exists or not in the database. And this check isn't asynchronous, will see how to execute it asynchronously.

**Comment by mokdennie26** [comment]
Thank you sir for your reply!

**Comment by MrTigreroux** [comment]
It should be fixed in new [version 4.12.11](https://www.spigotmc.org/resources/tigerreports.25773/update?update=421775). Please tell me if it is not the case. Have a good day.

**Comment by mokdennie26** [comment]
Hi sir. Thanks for the fix! But sadly I guess the issue is still there...

Uhmm I suggest the following:
If `ReportOnline: yes` (report online players ONLY)
Then when someone uses the report command `/report <name>`,
it checks whether the <name> parameter matches anyone who is currently online on the server (Arrays of name).
Maybe this can be done by storing names in an array / string list on join (and of course removing the names when they leave).
But I guess I've seen something similar in UserUtils. 

And, if, the <name> in the command does not match with any of the entries in the array, then maybe we shall skip checking whether that guy is in the database or not. That's because it's not necessary since `ReportOnline: yes`.

Because I believe the issue comes from that the plugin still looks for the offline guy in the database while `ReportOnline: yes`.
It would be a bit meaningless to do so.

The above will completely fix non-necessary resource consumption for those who set `ReportOnline: yes`.

But I guess the asynchronous database check may require a further fix for users with `ReportOnline: no`

**Comment by MrTigreroux** [comment]
Ok, there must be a problem in another place, will seek where.
The reason why the invalid check is performed before the offline check is that it is not logical to send an error telling that the provided user is offline when this user is invalid. If it is valid, then it is logical to send an error telling that it is offline.
And also the online check take in consideration the whole BungeeCord network, not only the current Spigot server.

**Comment by MrTigreroux** [comment]
And just to be sure, you say that the issue is still there, but are you observing a lag on the server or only a delay when you run the report command with an offline/invalid player ? Because the delay is normal (the plugin access to the database), the lag is anormal (the access should be asynchronously, it shouldn't disturb the whole server)

**Comment by mokdennie26** [comment]
Oh, it's a lag, sir. It's like everyone stuttered in the air for 0.5-1 second.

**Comment by MrTigreroux** [comment]
Should be fixed in new [version 4.12.12](https://www.spigotmc.org/resources/tigerreports.25773/update?update=422210). Make me know if it is still not the case.

**Comment by mokdennie26** [comment]
Thanks. it's fixed now.
